hashing/src/chunk_with_proof.rs

```diff
@@ -0,0 +1,304 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use casper_types::bytesrepr::{self, Bytes, FromBytes, ToBytes};
+
+use crate::{error::MerkleConstructionError, indexed_merkle_proof::IndexedMerkleProof, Digest};
+
+/// Represents a chunk of data with attached proof.
+#[derive(PartialEq, Debug, JsonSchema, Serialize, Deserialize)]
+#[schemars(with = "String", description = "Chunk of data with attached proof.")]
+#[serde(deny_unknown_fields)]
+pub struct ChunkWithProof {
+    proof: IndexedMerkleProof,
+    #[schemars(with = "String", description = "Hex-encoded bytes.")]
+    chunk: Bytes,
+}
+
+impl ToBytes for ChunkWithProof {
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut result = bytesrepr::allocate_buffer(self)?;
+        result.append(&mut self.proof.to_bytes()?);
+        result.append(&mut self.chunk.to_bytes()?);
+        Ok(result)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.proof.serialized_length() + self.chunk.serialized_length()
+    }
+}
+
+impl FromBytes for ChunkWithProof {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (proof, remainder) = FromBytes::from_bytes(bytes)?;
+        let (chunk, remainder) = FromBytes::from_bytes(remainder)?;
+
+        Ok((ChunkWithProof { proof, chunk }, remainder))
+    }
+}
+
+impl ChunkWithProof {
+    #[cfg(test)]
+    /// 10 bytes for testing purposes.
+    pub const CHUNK_SIZE_BYTES: usize = 10;
+
+    #[cfg(not(test))]
+    /// 1 MiB
+    pub const CHUNK_SIZE_BYTES: usize = 1 << 20;
```

Comment:
If the message size limit is 1 MiB, shouldn't we make the size limit somewhat smaller, so that at least `ChunkWithProof` (and probably even a bit more, as messages usually also contain some additional data) always fits in 1 MiB? Or is the actual limit a bit larger?

Reply:
By message size you mean the value configured by `maximum_net_message_size` in `chainspec.toml`? 

Anyway, if there is some interaction of the message size limit and chunk size that need to be taken care of, I'd suggest requesting a separate ticket.

@Fraser999 what's your opinion on this?

Reply:
IIRC (correct me if I'm wrong!) the main idea behind chunking was that we have an upper limit on the size of a single message in the networking layer, and that limit is 1 MiB. So when we have a trie node that is larger, we need to split it into smaller chunks, so that we'll be able to send it over the network. But if the limit is _exactly_ 1 MiB, then the chunks should be _smaller_, because the message will contain not only the chunk itself, but also some additional info (like the message variant tag, tag for the variant of `TrieOrChunkWithProof`, the Merkle proof for the chunk etc.), and all of this will have to fit within the message size limit.

Reply:
[Looks like it's actually configurable, and set to 22 MB?](https://github.com/casper-network/casper-node/blob/2be27b3f55af0ea43785117432aa26271636f95b/resources/production/chainspec.toml#L25) (Correct me if this isn't the (only) network message size limit!)

Anyway, if it's configurable and we hard-code the chunking to 1 MB, we should probably disallow configuration settings less than, say, 2 MB?

Reply:
Yes, that's why I was trying to confirm that we talk about the `maximum_net_message_size` in `chainspec.toml`.
Introducing a minimum limit for this param sounds like a good idea.

Reply:
#2354 should resolve this issue.